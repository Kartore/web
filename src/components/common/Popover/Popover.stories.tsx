import { useRef } from 'react';

import type { Meta, StoryObj } from '@storybook/react';
import { useOverlayTriggerState } from 'react-stately';
import { Popover } from '.';

const meta = {
	component: Popover,
} satisfies Meta<typeof Popover>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
	// @ts-expect-error render側でchildrenのみを受け取るため
	args: {
		children: 'test',
	},
	render: ({ children }) => {
		const ref = useRef(null);
		const state = useOverlayTriggerState({
			defaultOpen: true,
		});
		return (
			<>
				<button ref={ref} type={'button'}>
					Trigger
				</button>
				<Popover triggerRef={ref} state={state}>
					{children}
				</Popover>
			</>
		);
	},
};
