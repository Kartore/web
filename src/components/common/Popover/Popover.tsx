import type { FC, ReactNode } from 'react';
import { useRef } from 'react';

import type { AriaPopoverProps } from 'react-aria';
import { DismissButton, Overlay, usePopover } from 'react-aria';
import type { OverlayTriggerState } from 'react-stately';

export type PopoverProps = Omit<AriaPopoverProps, 'popoverRef'> & {
  children: ReactNode;
  state: OverlayTriggerState;
};

export const Popover: FC<PopoverProps> = ({ state, children, ...props }) => {
  const popoverRef = useRef(null);
  const { popoverProps, underlayProps } = usePopover(
    {
      ...props,
      popoverRef,
    },
    state
  );

  return (
    <Overlay>
      <div {...underlayProps} style={{ position: 'fixed', inset: 0 }} />
      <div
        {...popoverProps}
        ref={popoverRef}
        style={{
          ...popoverProps.style,
          background: 'var(--page-background)',
          border: '1px solid gray',
        }}
      >
        <DismissButton onDismiss={state.close} />
        {children}
        <DismissButton onDismiss={state.close} />
      </div>
    </Overlay>
  );
};
