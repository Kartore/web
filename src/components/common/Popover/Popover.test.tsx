import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { describe } from 'vitest';

import * as Stories from './Popover.stories';

const { Default } = composeStories(Stories);

describe('Component: Popover', () => {
	describe('Snapshot', () => {
		it('Default', () => {
			const { asFragment } = render(<Default />);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
