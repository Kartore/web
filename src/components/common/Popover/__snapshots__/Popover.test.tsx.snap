// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component: Popover > Snapshot > Default 1`] = `
<DocumentFragment>
  <button>
    Trigger
  </button>
</DocumentFragment>
`;
